pkg/kafka/encoding.go

```diff
@@ -0,0 +1,176 @@
+// Package kafka provides encoding and decoding functionality for Loki's Kafka integration.
+package kafka
+
+import (
+	"errors"
+	"fmt"
+	math_bits "math/bits"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/grafana/loki/v3/pkg/logql/syntax"
+)
+
+// Encoder is responsible for encoding logproto.Stream data into Kafka records.
+// It handles splitting of large streams into multiple records.
+type Encoder struct {
+	batch logproto.Stream
+}
+
+// NewEncoder creates a new Encoder.
+func NewEncoder() *Encoder {
+	return &Encoder{}
+}
+
+// Encode converts a logproto.Stream into one or more Kafka records.
+// It handles splitting large streams into multiple records if necessary.
+//
+// The encoding process works as follows:
+// 1. If the stream size is smaller than maxSize, it's encoded into a single record.
+// 2. For larger streams, it splits the entries into multiple batches, each under maxSize.
+// 3. The data is wrapped in a Kafka record with the tenant ID as the key.
+//
+// The format of each record is:
+// - Key: Tenant ID (used for routing, not for partitioning)
+// - Value: Protobuf serialized logproto.Stream
+// - Partition: As specified in the partitionID parameter
+//
+// Parameters:
+// - partitionID: The Kafka partition ID for the record
+// - tenantID: The tenant ID for the stream
+// - stream: The logproto.Stream to be encoded
+// - maxSize: The maximum size of each Kafka record
+func (e *Encoder) Encode(partitionID int32, tenantID string, stream logproto.Stream, maxSize int) ([]*kgo.Record, error) {
+	reqSize := stream.Size()
+
+	// Fast path for small requests
+	if reqSize <= maxSize {
+		rec, err := e.marshalWriteRequestToRecord(partitionID, tenantID, stream)
+		if err != nil {
+			return nil, err
+		}
+		return []*kgo.Record{rec}, nil
+	}
+
+	var records []*kgo.Record
+	e.batch.Labels = stream.Labels
+	e.batch.Hash = stream.Hash
+
+	if e.batch.Entries == nil {
+		e.batch.Entries = make([]logproto.Entry, 0, 1024)
+	}
+	e.batch.Entries = e.batch.Entries[:0]
+	labelsSize := e.batch.Size()
+	currentSize := labelsSize
+
+	for i, entry := range stream.Entries {
+		l := entry.Size()
+		// Size of the entry in the stream
+		entrySize := 1 + l + sovPush(uint64(l))
+
+		// Check if a single entry is too big
+		if entrySize > maxSize || (i == 0 && currentSize+entrySize > maxSize) {
+			return nil, fmt.Errorf("single entry size (%d) exceeds maximum allowed size (%d)", entrySize, maxSize)
+		}
+
+		if currentSize+entrySize > maxSize {
+			// Current stream is full, create a record and start a new stream
+			if len(e.batch.Entries) > 0 {
+				rec, err := e.marshalWriteRequestToRecord(partitionID, tenantID, e.batch)
+				if err != nil {
+					return nil, err
+				}
+				records = append(records, rec)
+			}
+			// Reset currentStream
+			e.batch.Entries = e.batch.Entries[:0]
```

Comment:
I'm not 100% sure I understand this. What is the purpose of resetting the current stream when `currentSize + entrySize > maxSize` and there are no entries in the batch?